### PR TITLE
use array pool instead of shared array

### PR DESCRIPTION
**Pull request by minggo**
shared array will have problem if dispatching an event in a event call back function and the event type is propagationStopped

Re: https://github.com/cocos/cocos-engine/issues/16914

### Changelog

*

-------

### Continuous Integration

This pull request:

* [ ] needs automatic test cases check.
  > Manual trigger with `@cocos-robot run test cases` afterward.
* [ ] does not change any runtime related code or build configuration
  > If any reviewer thinks the CI checks are needed, please uncheck this option, then close and reopen the issue.

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [ ] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

### CODE REVIEW

**Comment by minggo** [comment]
@cocos-robot run test cases

**Comment by sushanhong** [comment]
@cocos-robot run test cases

**Comment by sushanhong** [comment]
@cocos-robot run test cases

**Comment by sushanhong** [comment]
@cocos-robot run test cases

**Comment by sushanhong** [comment]
@cocos-robot run test cases

**Comment by sushanhong** [comment]
@cocos-robot run test cases

**Comment by sushanhong** [comment]
@cocos-robot run test cases

File: cocos/scene-graph/node-event-processor.ts [hunk 1]
```diff
@@ -291,24 +291,25 @@ export class NodeEventProcessor {
...
+            for (i = 0; i < cachedArray.length; ++i) {
+                target = cachedArray[i];
                 if (target.eventProcessor.bubblingTarget) {
                     event.currentTarget = target;
                     // fire event
                     target.eventProcessor.bubblingTarget.emit(event.type, event);
                     // check if propagation stopped
                     if (event.propagationStopped) {
-                        _cachedArray.length = 0;
                         return;
```
Comment (qiuguohua) [inline cocos/scene-graph/node-event-processor.ts:308]: Need to recycle cachedArray if it's returned here?

File: cocos/scene-graph/node-event-processor.ts [hunk 2]
```diff
@@ -259,24 +260,23 @@ export class NodeEventProcessor {
...
+            target = cachedArray[i];
             if (target.eventProcessor.capturingTarget) {
                 event.currentTarget = target;
                 // fire event
-                target.eventProcessor.capturingTarget.emit(event.type, event, _cachedArray);
+                target.eventProcessor.capturingTarget.emit(event.type, event, cachedArray);
                 // check if propagation stopped
                 if (event.propagationStopped) {
-                    _cachedArray.length = 0;
                     return;
```
Comment (qiuguohua) [inline cocos/scene-graph/node-event-processor.ts:277]: Need to recycle cachedArray if it's returned here?

**Comment by sushanhong** [comment]
@cocos-robot run test cases

**Comment by sushanhong** [comment]
@cocos-robot run test cases

**Comment by sushanhong** [comment]
@cocos-robot run test cases

**Comment by sushanhong** [comment]
@cocos-robot run test cases

File: cocos/scene-graph/node-event-processor.ts (hunk 2 above)
Comment (minggo) [reply cocos/scene-graph/node-event-processor.ts:277]: Oh, yep, need to free it.

File: cocos/scene-graph/node-event-processor.ts (hunk 1 above)
Comment (minggo) [reply cocos/scene-graph/node-event-processor.ts:308]: fixed

File: cocos/scene-graph/node-event-processor.ts (hunk 2 above)
Comment (minggo) [reply cocos/scene-graph/node-event-processor.ts:277]: fixed

**Review by qiuguohua — APPROVED** [review]

**Review by dumganhar — APPROVED** [review]

**Comment by minggo** [comment]
@cocos-robot run test cases

**Pull request merged**